### PR TITLE
:memo: Redesign the project `readme` page

### LINKED ISSUE #1150: :memo: Redesign the project `readme` page
#### Work Expected From The Issue
Redesign the project's `readme` page to make it look more proffessional, contributor friendly and welcoming.

The issue expects the following files to be changed/modified:
- `README.md`

> [!Note]
> All the files that are expected to be changed are located under the codebase (`xitca-web` directory).

#### Reason Behind These Changes
The reason behind having these changes is to redesign the `readme` page for the project so that it looks more proffesional, more clean, easier to read for new visitors and make it more welcoming and contributor friendly.

### Screenshots

_No response_

### Do you want to work on this issue?

Yes

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by neon-mmd**
## What does this PR do?

Redesigns the `readme` page of the project and also adds an image to be displayed on the page. 

## Why is this change important?

This change is essential as by redesigning the `readme` page it will make it more proffessional, cleaner, contributor-friendly and easier for new visitors to read and follow through.

## How to test this PR locally?

For a rendered version of the `readme` page. check:

https://github.com/neon-mmd/xitca-web/tree/DOCS/1150_redesign-the-projects-readme

## Author's checklist
- [x] Redesign the `readme` page.
- [x] Add a new image to be displayed after the project description.  

## Related issues

Closes #1150

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,15 +1,203 @@
...
+And then run the application/server using `cargo run` command and it should start the server/application on <http://127.0.0.1:8080>. Open your prefered browser and navigate to the address on which the server has started and you should see a `hello world!!` message appear.
+
+> [!Warning]
+> The project is still not stable. So expect breaking changes to appear at any time.
+
+**[⬆️ Back to Top](#xitca-web)**
+
+# FAQ (Frequently Asked Questions) 🗨️
+
+## Why Yet Another HTTP Framework?
```
Comment (fakeshadow) [inline README.md:111]: It will be better if this part explain the design motivation of xitca-web.(the points in previous readme)
It should also mention that `xitca-web` as a project is not only a web framework. It includes it's own http library, async IO abstraction, tls integration and even a database driver.

File: README.md [hunk 2]
```diff
@@ -1,15 +1,203 @@
...
+
+The primary reason behind creating Xitca-web is to have more memory safe HTTP framework which uses 100% safe rust unlike HTTP frameworks like actix-web which heavily rely upon unsafe rust code. As well as to provide a framework with low-memory footprint and with very small dependency tree.
```
Comment (fakeshadow) [inline README.md:113]: We should not mention any specific Rust crate in way that can possibly lead to negativity and contention. I feel we can just state xitca-web is 100% safe and remove comparison between actix-web.

File: README.md [hunk 3]
```diff
@@ -1,15 +1,203 @@
...
+  - [🙋 Contributing](#contributing-)
+  - [📜 License](#license-)
+  - [🤝 Credits](#credits-)
+
+  </p>
+</details>
+
+# Features 🚀
+
+- 🎗️ Supports HTTP/1.x and HTTP/2
```
Comment (fakeshadow) [inline README.md:52]: `xitca-web` project also supports HTTP/3 (with `xitca-http` specifically)

File: README.md [hunk 4]
```diff
@@ -1,15 +1,203 @@
...
+  <i>
+    A HTTP library and web framework written in 100% safe Rust with zero-copy
+    serialization/deserilization support
+  </i>
+</p>
 
-## Minimum Supported Rust Version
-- 1.79 [^1]
+<br />
+<img src="./images/server.jpg" alt="An image of a server" align="center" />
```
[the remark this answers is not in the document]
Comment (neon-mmd) [reply README.md:28]: > I would prefer this image removed. It serves no obvious purpose and can take up some extra resource when cloning the repo. In some part of the world accessing github can be hard and slow and I know this in first hand experience.

No I just kept this image because I have seen many popular projects do it. Also, I think images can attract the reader quite a lot and grabs their attention. So that's why. Additionally for speed I think we can compress the image down if you wish this can speed things by a lot. What do you say? :slightly_smiling_face:

File: README.md [hunk 5]
```diff
@@ -1,15 +1,203 @@
...
-- Lightweight dependency tree. Avoid adding unnecessary import when possible. Prefer no proc macro code generation when possible(proc macro feature are still offered as opt-in instead of opt-out).
-- Compact and simple code base. Reduce the barrier of understanding of source code for easier contributing.
-- Good interop with other crates. `tokio`(for async runtime) and `http`(for http types) are used directly as dependency.
+<details>
+  <summary><b>Table of Contents</b></summary>
+  <p>
 
-[^1]: `http-file` excluded due to design issue.
-[^2]: only guaranteed inside project's own code. unsafe can and is used by dependencies
+  - [🚀 Features](#features-)
```
[the remark this answers is not in the document]
Comment (neon-mmd) [reply README.md:35]: > I believe we should not include emojis in the readme. They are fun but can be seen as less serious. Boring and plain is preferred.

No I believe emojis can grab the user visually which is great as well. Also, for this I would suggest you can go through many popular projects on the trending they do this quite often. So I don't think it hurts in anyway. What do you think? :slightly_smiling_face: .

File: README.md [hunk 6]
```diff
@@ -1,15 +1,203 @@
...
+
+To get started using Xitca-web add the following code to your `main.rs` file: 
+```rust
+use xitca_web::{handler::handler_service, middleware::Logger, route::get, App};
+
+async fn index() -> &'static str {
+    "Hello world!!"
+}
+
+#[tokio::main]
```
[the remark this answers is not in the document]
Comment (neon-mmd) [reply README.md:83]: > We should remove the tokio main macro here to high light the macroless default of `xitca-web`. It's not only the web framework itself but the whole user dependency tree can be proc macro crates free.

Yes I think I agree with you. So I will make the change as soon as possible :slightly_smiling_face: .

File: README.md (hunk 1 above)
Comment (neon-mmd) [reply README.md:111]: > It will be better if this part explain the design motivation of xitca-web.(the points in previous readme) It should also mention that `xitca-web` as a project is not only a web framework. It includes it's own http library, async IO abstraction, tls integration and even a database driver.

Ok I see, I didn't quite knew about them but I will mention these in this section. :slightly_smiling_face:

File: README.md (hunk 2 above)
Comment (neon-mmd) [reply README.md:113]: > We should not mention any specific Rust crate in way that can possibly lead to negativity and contention. I feel we can just state xitca-web is 100% safe and remove comparison between actix-web.

I don't think we are making a direct comparison here like if you read through it I am mentioning more like in general comparison to other projects and if you read carefully you will find that `actix-web` is just used as kind of an example. :sweat_smile: . But if you still suggest having it removed then I would not mind doing so. :slightly_smiling_face:

File: README.md (hunk 3 above)
Comment (neon-mmd) [reply README.md:52]: > `xitca-web` project also supports HTTP/3 (with `xitca-http` specifically)

Ok sure, I will mention about this also in the `features` section. :slightly_smiling_face:

File: README.md (hunk 2 above)
Comment (fakeshadow) [reply README.md:113]: It's not really about comparison but the possibility of misinterpretation and misunderstanding. For example a reader may form a negative view towards the project from your statement for following reason:
1. `actix-web` does not rely heavily on unsafe. It's code base has less than a dozen direct usage of `unsafe` key word. That's a very low amount of usage judging by it's code base.
2. `actix-web` has it's unsafe usage audited and there is no apparent soundness issue.
3. `usafe` is a language feature. Using it or not is based on design choice, trade off and execution. In most use cases it's an opinion of people and there could be no objectively correct answer (if unsafe should be used or not)

In other words it's very hard to make fair comparison and word it correctly in a short readme. And the best thing to do is avoid making comparison when possible. Library users are smart enough to figure things out themselves and form their own opinion.

File: README.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (fakeshadow) [reply README.md:35]: I get what you mean and I appreciate the thought you put into it. 

That said `xitca-web` is not aiming for popularity and we have zero interest in "hyper" or "trend". Being plain and boring is an explicit choice from the start for the tone of the project.

File: README.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (fakeshadow) [reply README.md:28]: I would still prefer leaving anything that are not functionally part of the project out. In some countries accessing github can be as slow as single digit kilobytes per second. An irrelevant image even compressed can waste people's valuable time and bandwidth.

There is also the issue that irrelevant stuff can pile up over time if we accept the practice. There is always more asset can be included just for the look.

File: README.md (hunk 2 above)
Comment (neon-mmd) [reply README.md:113]: > It's not really about comparison but the possibility of misinterpretation and misunderstanding. For example a reader may form a negative view towards the project from your statement for following reason:
> 
> 1. `actix-web` does not rely heavily on unsafe. It's code base has less than a dozen direct usage of `unsafe` key word. That's a very low amount of usage judging by it's code base.
> 2. `actix-web` has it's unsafe usage audited and there is no apparent soundness issue.
> 3. `usafe` is a language feature. Using it or not is based on design choice, trade off and execution. In most use cases it's an opinion of people and there could be no objectively correct answer (if unsafe should be used or not)
> 
> In other words it's very hard to make fair comparison and word it correctly in a short readme. And the best thing to do is avoid making comparison when possible. Library users are smart enough to figure things out themselves and form their own opinion.

Ah ok I see, you are right. Ok then I will just reframe this section to not compare other projects in anyway. Thanks :heart: for your time explaining everything in depth. :slightly_smiling_face:

File: README.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (neon-mmd) [reply README.md:35]: > I get what you mean and I appreciate the thought you put into it.
> 
> That said `xitca-web` is not aiming for popularity and we have zero interest in "hyper" or "trend". Being plain and boring is an explicit choice from the start for the tone of the project.

Also, I will point out that having emojis in this secton also doesn't matter because this will fold up anyway so if will remain hidden. Additionally I am not aiming at trending though but I feel having this can grab the reader's attention and also attracts contribution like it gives the impression like `wow such a cool project should try it out or should contribute to it`. But anyways since you insist I will remove it. :slightly_smiling_face:

File: README.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (neon-mmd) [reply README.md:28]: > I would still prefer leaving anything that are not functionally part of the project out. In some countries accessing github can be as slow as single digit kilobytes per second. An irrelevant image even compressed can waste people's valuable time and bandwidth.
> 
> There is also the issue that irrelevant stuff can pile up over time if we accept the practice. There is always more asset can be included just for the look.

I don't think it should be a major problem I mean even if the image does not load I think it is still readable though but I think having a blank readme again is quite bad like it is does not grab the reader's attention alot and it could make the project left out of significant users or contributors I feel. But again if you insist I will remove it :slightly_smiling_face: .0

**Review by fakeshadow — APPROVED** [review]
LGTM. Thanks for PR
